### ISSUE TITLE
Error Date

### ISSUE THREAD

**Issue by ramazancivak** [nager/Nager.Date#611]
### Affected country

Türkiye 

### Affected public holiday

Eid al-Fitr First 

### Source of the information

https://tr.wikipedia.org/wiki/Ramazan_Bayram%C4%B1

### More information

https://date.nager.at/Api/v2/PublicHolidays/2024/TR 

2024.03.30 instead of 2024.04.10

**Comment by tinohager** [comment]
please take note of the information [WebApi V1 and V2 deprecated](https://github.com/nager/Nager.Date/issues/575)

**Comment by LayerCakeMakes** [comment]
This is wrong in WebApi V3 too for all three days.
These are correct: https://www.timeanddate.de/feiertage/tuerkei/
These are not (returned by V3):
```
  {
    "date": "2024-03-30",
    "localName": "Ramazan Bayramı 1. Gün",
    "name": "Eid al-Fitr First Day",
    "countryCode": "TR",
    "fixed": false,
    "global": true,
    "counties": null,
    "launchYear": null,
    "types": [
      "Public"
    ]
  },
  {
    "date": "2024-03-31",
    "localName": "Ramazan Bayramı 2. Gün",
    "name": "Eid al-Fitr Second Day",
    "countryCode": "TR",
    "fixed": false,
    "global": true,
    "counties": null,
    "launchYear": null,
    "types": [
      "Public"
    ]
  },
  {
    "date": "2024-04-01",
    "localName": "Ramazan Bayramı 3. Gün",
    "name": "Eid al-Fitr Third Day",
    "countryCode": "TR",
    "fixed": false,
    "global": true,
    "counties": null,
    "launchYear": null,
    "types": [
      "Public"
    ]
  },
```

**Comment by tinohager** [comment]
Where can I find out the exact date for Ramadan? It is set every year by a church leader. Is there a government source here?

**Comment by LayerCakeMakes** [comment]
As far as I understand it, it is roughly based on a lunar calendar but it depends on when the cresent moon is actually visible/seen so it's a tentative date until just a few days before the actual date when it will be announced to the public, but who announces it varies from country to country and not all countries might have it start on the exact same date.

**Comment by tinohager** [comment]
I am aware of that. But where do I get the real date? And does the date only apply to Turkey or also to other countries?

**Comment by LayerCakeMakes** [comment]
It's different for each country. For turkey it seems to be Directorate of Religious Affairs also known as Diyanet. From the articels I've found it seems they announced it on Instagram??? Unfortunately I don't speak the language so I couldn't verify that.

**Comment by LayerCakeMakes** [comment]
Maybe for V4 these kinds of holidays need a marker for tentative dates that might still change.

I just noticed that your calculation for the tentative date you supply seems to be off by 1 year.
The tentative date for 2024 is 2024-04-01 but the api returned 2024-03-30 which matches the tentative date for 2025 which is 2025-03-30. This seem to continue on to other future years.

**Comment by LayerCakeMakes** [comment]
There is another issue that might not exactly belong in this report but it is related. Some holidays in Turkey start half a day early at 13:00 like Eid al-Fitr First, so this might be a marker in V4 too or half day support.

**Comment by tinohager** [comment]
I have temporarily deactivated the faulty holidays. Until there is a sensible solution for the future.

**Comment by hasankaplan-github** [comment]
According to [Wikipedia](https://en.wikipedia.org/wiki/Hijri_year) one single Gregorian year can match with three Hijri years. So my solution is returning multiple Hijri years instead of one Hijri year (calculated Hijri year, minus 1, and plus 1.) 

```
private int[] GetPossibleHijriYears(int year)
{
    var diff = year - 621;
    var hijriYear = Convert.ToInt32(Math.Round(diff + decimal.Divide(diff, 33)));
    return [hijriYear - 1, hijriYear, hijriYear + 1];
}
```

Then calculate multiple holidays according to multiple Hijri years and just select the dates which has correct year.

```
public IEnumerable<SpecialDay> RamadanFirstDay(int year)
{
    var hijriCalendar = new UmAlQuraCalendar();
    var possibleHijriYears = GetPossibleHijriYears(year);

    foreach (var hijriYear in possibleHijriYears)
    {
        var dateTime = hijriCalendar.ToDateTime(hijriYear, 10, 1, 0, 0, 0, 0);
        if (dateTime.Year != year)
        {
            continue;
        }

        var date = DateOnly.FromDateTime(dateTime);

        yield return new SpecialDay(
            id: Guid.Empty,
            isHoliday: true,
            date: date,
            name: "Ramazan Bayramı 1. Gün",
            isFixed: false);
    }
}
```

**Comment by hasankaplan-github** [comment]
I want to update my previous `GetPossibleHijriYears` method. I think `GetHijriYears` method below is more accurate then the previous `GetPossibleHijriYears` method:

```
private int[] GetHijriYears(int year)
{
    var hijriCalendar = new UmAlQuraCalendar();

    var minHijriYear = hijriCalendar.GetYear(new DateTime(year, 1, 1));
    var maxHijriYear = hijriCalendar.GetYear(new DateTime(year, 12, 31));

    var hijriYears = Enumerable.Range(minHijriYear, maxHijriYear - minHijriYear + 1).ToArray();

    return hijriYears;
}
```

**Comment by tinohager** [comment]
@ramazancivak @hasankaplan-github @LayerCakeMakes 
Can any of you provide reliable holiday dates for the past few years?

**Comment by LayerCakeMakes** [comment]
@tinohager https://www.timeanddate.de/feiertage/tuerkei/2025 seems to have good history and estimated future date.

The main problem with the estimated future dates was that you can't get an accurate hijiri year from just a gregorian year you need the full gregorian date to see which hijiri year it is as there can be three different hijiri years during one gregorian year.

**Comment by tinohager** [comment]
Here is an export of the modified logic, which should be pretty much correct.

|TR-EIDALFITRFIRSTDAY-01|
|-----------------------|
|2006-10-23             |
|2007-10-12             |
|2008-10-01             |
|2009-09-20             |
|2010-09-08             |
|2011-08-29             |
|2012-08-17             |
|2013-08-06             |
|2014-07-27             |
|2015-07-16             |
|2016-07-05             |
|2017-06-25             |
|2018-06-14             |
|2019-06-04             |
|2020-05-22             |
|2021-05-11             |
|2022-05-01             |
|2023-04-21             |
|2024-04-09             |
|2025-03-29             |
|2026-03-20             |
|2027-03-08             |
|2028-02-25             |
|2029-02-13             |
|2030-02-03             |
|2031-01-24             |
|2032-01-13             |
|2033-01-02             |
|2034-12-11             |
|2035-12-01             |
|2036-11-19             |
|2037-11-09             |
|2038-10-29             |
|2039-10-18             |
|2040-10-07             |
|2041-09-26             |
|2042-09-15             |
|2043-09-05             |
|2044-08-24             |
|2045-08-14             |
|2046-08-03             |

**Comment by hasankaplan-github** [comment]
I did not even aware of "adjustments" for Hijri date calculations for Turkiye. Your work is respectable. I have viewed your "turkey-hijri" branch. One point to say is "GetEidAlFitr" and "GetEidAlAdha" methods return early without calculating second occurance of holidays within a single gregorian year (one possible at the beginning of the year and nearly at the end of the year.) For this reason maybe "HolidaySpecification"s should be put in a collection and then should be return that collection. I mean "return" should be outside of the "for" loop, not in the "for" loop.
Thank you.

**Comment by tinohager** [comment]
Yes, you're right—it should definitely go through the loop and collect everything into a collection, rather than stopping right away.

**Comment by hasankaplan-github** [comment]
Where do we find explanations about "Hijri Adjustments"? Where did you find them?

**Comment by tinohager** [comment]
@hasankaplan-github I still need to adjust the list so that the correct date for each year is correct. In the future, we'll need to fine-tune the settings for each year here.

**Comment by tinohager** [comment]
The new logic has been published - could you please take a look at it?

**Comment by tinohager** [comment]
@hasankaplan-github I would really appreciate any feedback.

**Comment by hasankaplan-github** [comment]
> [@hasankaplan-github](https://github.com/hasankaplan-github) I would really appreciate any feedback.
@tinohager 

Hi, I calculate these correct adjustments from Diyanet web site: (please recheck them)
(1449, 10) -> 0
(1450, 10) -> 0
(1451, 10) -> 0
(1451, 12) -> 0
(1452, 10) -> 0
(1452, 12) -> 0
(1453, 10) -> -1
(1453, 12) -> -1
(1454, 10) -> 0
(1454, 12) -> 0
(1455, 10) -> -1
(1455, 12) -> -1
(1456, 10) -> -1
(1456, 12) -> -1
(1457, 10) -> 0
